pkg/kube/rbac/policy_rule_index.go

```diff
@@ -0,0 +1,171 @@
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+const (
+	All       = "*"
+)
+
+type policyRuleIndex struct {
+	crCache             v1.ClusterRoleCache
+	rCache              v1.RoleCache
+	crbCache            v1.ClusterRoleBindingCache
+	rbCache             v1.RoleBindingCache
+	revisions           *roleRevisionIndex
+	kind                string
+	roleIndexKey        string
+	clusterRoleIndexKey string
+}
+
+func newSAPolicyRuleIndex(revisions *roleRevisionIndex, rbac v1.Interface) *policyRuleIndex {
+	key := "ServiceAccount"
+
+	pi := &policyRuleIndex{
+		kind:                key,
+		crCache:             rbac.ClusterRole().Cache(),
+		rCache:              rbac.Role().Cache(),
+		crbCache:            rbac.ClusterRoleBinding().Cache(),
+		rbCache:             rbac.RoleBinding().Cache(),
+		clusterRoleIndexKey: "crb" + key,
+		roleIndexKey:        "rb" + key,
+		revisions:           revisions,
+	}
+
+	pi.crbCache.AddIndexer(pi.clusterRoleIndexKey, pi.clusterRoleBindingBySubjectIndexer)
+	pi.rbCache.AddIndexer(pi.roleIndexKey, pi.roleBindingBySubject)
+
+	return pi
+}
+
+func (p *policyRuleIndex) clusterRoleBindingBySubjectIndexer(crb *rbacv1.ClusterRoleBinding) (result []string, err error) {
+	for _, subject := range crb.Subjects {
+		if subject.Kind == p.kind && crb.RoleRef.Kind == "ClusterRole" {
+			result = append(result, subject.Name)
+		}
+	}
+	return
+}
+
+func (p *policyRuleIndex) roleBindingBySubject(rb *rbacv1.RoleBinding) (result []string, err error) {
+	for _, subject := range rb.Subjects {
+		if subject.Kind == p.kind {
+			result = append(result, subject.Name)
+		}
+	}
+	return
+}
+
+var null = []byte{'\x00'}
+
+func (p *policyRuleIndex) addRolesToHash(digest hash.Hash, subjectName string) {
+	for _, crb := range p.getClusterRoleBindings(subjectName) {
+		digest.Write([]byte(crb.RoleRef.Name))
+		digest.Write([]byte(p.revisions.roleRevision("", crb.RoleRef.Name)))
+		digest.Write(null)
+	}
+
+	for _, rb := range p.getRoleBindings(subjectName) {
+		switch rb.RoleRef.Kind {
+		case "Role":
+			digest.Write([]byte(rb.RoleRef.Name))
+			digest.Write([]byte(rb.Namespace))
+			digest.Write([]byte(p.revisions.roleRevision(rb.Namespace, rb.RoleRef.Name)))
+			digest.Write(null)
+		case "ClusterRole":
+			digest.Write([]byte(rb.RoleRef.Name))
+			digest.Write([]byte(rb.Namespace))
+			digest.Write([]byte(p.revisions.roleRevision(rb.Namespace, rb.RoleRef.Name)))
+			digest.Write(null)
+		}
+	}
+}
+
+func (p *policyRuleIndex) get(subjectName string) *accesscontrol.AccessSet {
+	result := &accesscontrol.AccessSet{}
+
+	for _, binding := range p.getRoleBindings(subjectName) {
+		p.addAccess(result, binding.Namespace, binding.RoleRef)
+	}
+
+	for _, binding := range p.getClusterRoleBindings(subjectName) {
+		p.addAccess(result, All, binding.RoleRef)
+	}
+
+	return result
+}
+
+func (p *policyRuleIndex) addAccess(accessSet *accesscontrol.AccessSet, namespace string, roleRef rbacv1.RoleRef) {
```

Comment:
There are five `for` blocks in its implementation, can we separate them?

Reply:
updated